proposals/text/5540-llm-support-2.0.md

```diff
@@ -0,0 +1,309 @@
+
+pipe.connect("retriever", "prompt")
+pipe.connect("prompt", "llm")
+pipe.connect("llm", "replies_converter")
+
+questions = ["Why?", "Why not?"]
+results = pipe.run({
+	"retriever": {"queries": questions},
+	"prompt": {"questions": questions},
+})
+
+assert results == {
+	"replies_converter": {
+    "answers": [[Answer("Because of this.")], [Answer("Because of that.")]]
+  }
+}
+```
+
+# Detailed design
+
+Haystack’s `PromptNode` is a very complex component that includes under its name several functionalities: loading prompt templates through the `PromptTemplate` class, rendering such prompt template with the variables from the invocation context, choosing which LLM backend to use, sending the prompt to the LLM using the correct invocation layer, interpreting the results, parsing them into objects, and putting them back in the pipeline in a way other components can understand.
+
+in Haystack 2.0 we unpack these functionalities into a few separate components, to clarify what is happening, how it works, and provide additional flexibility.
+
+The main functionalities we identified are the following:
+
+1. Fetching the prompt from different sources
+2. Rendering the prompt using variables
+3. Invoke the LLM
+4. Parse the output
+5. History/Memory management
+
+We leave the discussion about History/Memory to a separate proposal, as it concerns mostly Agents, and focus on the other points.
+
+## LLM invocation
+
+In Haystack 1.x, `PromptNode` uses `InvocationLayer` to query different LLMs under a unified API. In that design, users do not need to know which invocation layer is used for the model they select, as `PromptNode` takes responsibility of selecting it.
+
+Such invocation layers can be ported to 2.0 as standalone components. In this way we will have one component for each LLM backed that we support.
+
+Each component should be named after the class of models it supports. For example we should have `GPT4`, `ChatGPT`, etc. For components whose name may be confusing, or that include models that are not LLMs, we can add a LLM suffix: like `FalconLLM`, `CohereLLM`, `HuggingFaceLLM`.
+
+Note that having separate components for each LLM makes easy to deprecate them when we realize they are dropping out of favor or become severely outdated. It also makes very easy for external contributors to make their own external components to support rarer LLMs, without having to add them to Haystack’s core.
+
+All these LLM clients will have a near-identical I/O:
+
+```python
+@component
+class ChatGPT:
+
+    @component.output_types(replies=List[List[str]])
+    def run(self, prompts: List[str], ... chatgpt specific params...):
+        ...
+        return {'replies': [...]}
+```
+
+Note how the component takes a list of prompts and LLM parameters only, but no variables nor templates, and returns only strings. This is because input rendering and output parsing are delegated to separate components, which description follows.
```

Comment:
What's the list of prompts for? When it should generate answers for multiple documents? How dose the chat interface work (system, user, AI message)?

Reply:
Currently in the PromptHub we have some prompt templates that generate more than one prompt when rendered, for example https://github.com/deepset-ai/prompthub/blob/main/prompts/deepset-question-answering-per-document.md This means that the `Prompts` component may output more than one string after a single call. That's why the input of the LLMs is a list of prompts: so that we won't have trouble connecting them. This also enables passing several queries together for other reasons, like batching of queries on the user's end.

Reply:
Regarding the chat interface, I purposefully didn't take it into account here. Chat system and Agents are for another proposal, so let's postpone the discussion for now

Reply:
> Currently in the PromptHub we have some prompt templates that generate more than one prompt when rendered, for example https://github.com/deepset-ai/prompthub/blob/main/prompts/deepset-question-answering-per-document.md This means that the `Prompts` component may output more than one string after a single call. That's why the input of the LLMs is a list of prompts: so that we won't have trouble connecting them. This also enables passing several queries together for other reasons, like batching of queries on the user's end.

This doesn't sound correct to me. The prompt you linked is a single prompt and it returns a string as any other prompt. In this case is the `AnswerParser` that makes it return a list of `Answer`s, there's only a single prompt.

> Regarding the chat interface, I purposefully didn't take it into account here. Chat system and Agents are for another proposal, so let's postpone the discussion for now

I think Mathis is talking about ChatGPT likes. Their [Chat completion](https://platform.openai.com/docs/api-reference/chat) from the ["classic" completion](https://platform.openai.com/docs/api-reference/completions).

I think it's important that we support them as other LLMs, building inputs for these kind of components will be a bit different though as they won't need just a string prompt.

Reply:
I understand that the memory and history will be accounted for Agents in a separate proposal. 

However, already in the PromptNode (for the RAG use case) a lot of these models depend on the passage of a separate system and user prompt to be used correctly. There is an open issue for this in Haystack currently https://github.com/deepset-ai/haystack/issues/5442 

I think it would be important to support this separation for the models that require it (e.g. GPT4, ChatGPT, LLama-2, etc.).

Reply:
Just to clarify @silvanocerza's comment: the mechanism by which PromptNode multiplies a single prompt template into several rendered prompts does not depend on the output parser, but on the template itself

<details><summary>Here is an example in action</summary>

```python

from haystack.nodes import PromptNode, PromptTemplate
from haystack.schema import Document

docs = [Document("The main suspect is John"), Document("The main suspect is Jane")]
prompt_template = PromptTemplate("deepset/question-answering-per-document")

prompt_node = PromptNode(model_name_or_path="text-davinci-003", api_key="...")
results = prompt_node.prompt(prompt_template=prompt_template, query="Who is the main suspect?", documents=docs)

print(results)

# ['John.', 'Jane is the main suspect.']
```

</details>

Reply:
Errata corrige on what I said above.

It's not the `AnswerParser` that makes it return two responses, it's the `PromptTemplate` class that renders the received string template to different prompts depending on the number of documents received. Then the `PromptNode` separately, that's why we receive multiple answers.

In my opinion this is confusing and not clear at all.

Given a prompt string template to a renderer you must always receive back a single prompt and not multiple ones.

Reply:
What @ZanSara said is what I meant by "it should generate answers for multiple documents". Understood then.

@silvanocerza this is needed, sometimes you might want to generate an answer for each document. The use case becomes more obvious if we change answer to summary. So if you want to generate summaries, you'd like to generate a summary for each document. How would you propose doing this if `Prompts` should only return one prompt?

Reply:
That's counter-intuitive for me, I would prefer if we find a better and clearer way of handling this use case.

Reply:
> However, already in the PromptNode (for the RAG use case) a lot of these models depend on the passage of a separate system and user prompt to be used correctly. There is an open issue for this in Haystack currently https://github.com/deepset-ai/haystack/issues/5442

Yes that's totally reasonable. I think we will end up having a non-chat version of these LLMs for RAG, and a chat version of them for more Agent-like workflows.  Would that work? If we do so, I believe the non-chat versions of the models would fit the current design: for example we could have the system prompt as an init parameter.

Reply:
> system prompt as an init parameter

I think this would work well!

Reply:
Note that not all chat systems need to be agents.

Reply:
@mathislucka you bring up a nice point because we've been discussing also what the definition of an Agent really is, and chat system are borderline, even though we tend to lean towards considering them very simple Agents. 

Let's anyway keep this discussion of what's an Agent and what isn't for the proposal about Agents, which I promise is in the works too. I'll soon point you to the Notion page where we can iterate over these points, let's not do it in this thread.

Reply:
I modified the proposal to clarify a bit what drives the input type of LLM components. I'll resolve this thread for now and reference it in a new comment.